MarkupsToModel/Logic/vtkSlicerMarkupsToModelLogic.cxx

```diff
@@ -292,45 +296,98 @@ void vtkSlicerMarkupsToModelLogic::UpdateOutputCloseSurfaceModel(vtkMRMLMarkupsT
+    coords[i][0] = markupPoint[0];
+    coords[i][1] = markupPoint[1];
+    coords[i][2] = markupPoint[2];
     modelCellArray->InsertCellPoint(i);
   }
+  vtkMath::MultiplyScalar(meanPoint, 1.0 / numberOfMarkups);
+
+  double corner[3] = {0.0, 0.0, 0.0};
+  double normal[3] = {0.0, 0.0, 0.0};
+  double temp[3] = {0.0, 0.0, 0.0};
+  vtkSmartPointer<vtkOBBTree> obbTree = vtkSmartPointer<vtkOBBTree>::New();
+  obbTree->ComputeOBB(modelPoints, corner, temp, temp, normal, temp);
+
+  double relative[3] = {0.0, 0.0, 0.0};
+  vtkMath::Subtract(meanPoint, corner, relative);
+  vtkMath::Normalize(normal);
+  double proj[3];
+  std::copy(normal, normal+3, proj);
+  double dotProd = vtkMath::Dot(relative, normal);
+  vtkMath::MultiplyScalar(proj, dotProd);
+
+  double origin[3];
+  vtkMath::Add(corner, proj, origin);
+
+  double A = normal[0];
+  double B = normal[1];
+  double C = normal[2];
+  double x0 = origin[0];
+  double y0 = origin[1];
+  double z0 = origin[2];
+  double D = (-1)*A*x0 - B*y0 - C*z0;
+
+  for (int i = 0; i < numberOfMarkups; i++)
+  {
+    double x1 = coords[i][0];
+    double y1 = coords[i][1];
+    double z1 = coords[i][2];
+    double distance = std::abs(A*x1 + B*y1 + C*z1 + D) / std::sqrt(A*A + B*B + C*C);
+    if (distance >= MINIMUM_THICKNESS)
+    {
+      OpenSurface = false;
+    }
+  }
 
   vtkSmartPointer< vtkPolyData > pointPolyData = vtkSmartPointer< vtkPolyData >::New();
   pointPolyData->SetLines(modelCellArray);
   pointPolyData->SetPoints(modelPoints);
 
   vtkSmartPointer< vtkDelaunay3D > delaunay = vtkSmartPointer< vtkDelaunay3D >::New();
   delaunay->SetAlpha(markupsToModelModuleNode->GetDelaunayAlpha());
-  if(markupsToModelModuleNode->GetCleanMarkups())
+
+  if (OpenSurface)
   {
-    vtkSmartPointer< vtkCleanPolyData > cleanPointPolyData = vtkSmartPointer< vtkCleanPolyData >::New();
-#if (VTK_MAJOR_VERSION <= 5)
-    cleanPointPolyData->SetInput(pointPolyData);
-#else
-    cleanPointPolyData->SetInputData(pointPolyData);
-#endif
-;
-    cleanPointPolyData->SetTolerance(CLEAN_POLYDATA_TOLERANCE);
-    delaunay->SetInputConnection(cleanPointPolyData->GetOutputPort()); //TODO SET VTK5
+    vtkSmartPointer<vtkCubeSource> cube = vtkSmartPointer<vtkCubeSource>::New();
+    vtkSmartPointer<vtkGlyph3D> glyph = vtkSmartPointer<vtkGlyph3D>::New();
+    glyph->SetSourceConnection(cube->GetOutputPort());
+    glyph->SetInputData(pointPolyData);
+    glyph->Update();
+    delaunay->SetInputConnection(glyph->GetOutputPort());
   }
   else
   {
+    if(markupsToModelModuleNode->GetCleanMarkups())
+    {
+      vtkSmartPointer< vtkCleanPolyData > cleanPointPolyData = vtkSmartPointer< vtkCleanPolyData >::New();
 #if (VTK_MAJOR_VERSION <= 5)
-    delaunay->SetInput( pointPolyData );
+      cleanPointPolyData->SetInput(pointPolyData);
 #else
-    delaunay->SetInputData(pointPolyData);
+      cleanPointPolyData->SetInputData(pointPolyData);
 #endif
+      cleanPointPolyData->SetTolerance(CLEAN_POLYDATA_TOLERANCE);
```

Comment:
It would be nice to keep this module generic, so that it works reasonably for objects that are 1m, 1mm, 1 um. Size of the point cloud just gives an approximate scale.
I think point placement error may be somewhat fixed in screen coordinates, so it may be an option to get the scale from the viewer's zoom factor, but I think it may be simpler to just use the point cloud extent to get an approximate scale.